vllm_spyre/v1/worker/spyre_model_runner.py

```diff
@@ -2180,43 +2212,50 @@ def _maybe_load_prefix_from_cache(self, scheduler_request: Request) -> int:
-                computed_blocks: list[
-                    KVCacheBlock] = FullAttentionManager.find_longest_cache_hit(
-                        block_hashes=scheduler_request.block_hashes,
-                        max_length=max_cache_hit_length,
-                        kv_cache_group_ids=[0],
-                        block_pool=self.block_pool,
-                        kv_cache_spec=self._attn_spec,
-                        use_eagle=False,
-                        dcp_world_size=1,
-                        **self._alignment_token_kwargs,
-                    )[0]
-                n_hit = len(computed_blocks)
-            else:
-                computed_blocks = list[KVCacheBlock]()
-                n_hit = 0
+
+            computed_blocks: list[
+                KVCacheBlock] = FullAttentionManager.find_longest_cache_hit(
+                    block_hashes=scheduler_request.block_hashes,
+                    max_length=prompt_len,
+                    kv_cache_group_ids=[0],
+                    block_pool=self.block_pool,
+                    kv_cache_spec=self._attn_spec,
+                    use_eagle=False,
+                    dcp_world_size=1,
+                    **self._alignment_token_kwargs,
+                )[0]
+            n_hit = len(computed_blocks)
+
             logger.debug("Found: %d cached_blocks", n_hit)
 
-            # trim down to chunk boundary
-            usable_blocks = (((left_blocks + n_hit) // self.chunk_blocks_count)\
-                * self.chunk_blocks_count) - left_blocks
-            usable_blocks = max(usable_blocks, 0)
-            logger.debug("Found: %d usable blocks in cache", usable_blocks)
-            computed_blocks = computed_blocks[:usable_blocks]
+            full_chunks_with_cached_blocks = ((left_blocks + n_hit) //
+                                              self.chunk_blocks_count)
+
+            # the last chunk of the prompt must always be recomputed
+            if full_chunks_with_cached_blocks == chunk_count:
+                full_chunks_with_cached_blocks -= 1
+
+            usable_blocks = max(
+                0, (full_chunks_with_cached_blocks * self.chunk_blocks_count) -
+                left_blocks)
+
+            # blocks to compute from scratch or recompute in the last chunk
+            blocks_to_compute = padded_prompt_len // self.block_size \
+                - usable_blocks
+            logger.debug(
+                "Found: %d reusable blocks in cache. "
+                "%d blocks will be (re)computed", usable_blocks,
+                blocks_to_compute)
             num_cached_tokens = usable_blocks * self.block_size
 
+            # Save all of the computed blocks and not only the usable
+            # ones because we will make a dummy recomputation of computed
+            # blocks in the last chunk to deduplicate the used blocks. So
+            # although we will recompute, we'll still point the block table
+            # to the cached blocks.
```

Comment:
Since we already know `computed_blocks`, `usable_blocks`, and `blocks_to_compute` here, could we save some of this info in the prefill batch for this request instead of requiring the logic to recalculate the number of recomputed blocks in the next pass?

Reply:
That's not a bad idea. We can subclass the SamplingInputBatch to add these extra attributes and add builder method to the cb model runner that we can override in the cp model runner. What if we do this as a follow-up so that we can get the linter migration PR in?